lib/kafka/multi_subscription_round_robin_assignment_strategy.rb

```diff
@@ -0,0 +1,56 @@
+module Kafka
+
+  # A round robin assignment strategy suitable for when subscriptions
+  # differ across consumer instances in the same consumer group.
+  class MultiSubscriptionRoundRobinAssignmentStrategy
+    def protocol_name
+      "roundrobin"
+    end
+
+    # Assign the topic partitions to the group members.
+    #
+    # @param cluster [Kafka::Cluster]
+    # @param members [Hash<String, Kafka::Protocol::JoinGroupResponse::Metadata>] a hash
+    #   mapping member ids to metadata
+    # @param partitions [Array<Kafka::ConsumerGroup::Assignor::Partition>] a list of
+    #   partitions the consumer group processes
+    # @return [Hash<String, Array<Kafka::ConsumerGroup::Assignor::Partition>] a hash
+    #   mapping member ids to partitions.
+    def call(cluster:, members:, partitions:)
+      partitions_per_member = Hash.new {|h, k| h[k] = [] }
+      relevant_partitions = valid_sorted_partitions(members, partitions)
+      members_ids = members.keys
+      idx = 0
+
+      relevant_partitions.each do |partition|
+        topic = partition.topic
+
+        while !members[members_ids[idx]].topics.include?(topic)
+          idx = next_index(members_ids, idx)
+        end
+
+        partitions_per_member[members_ids[idx]] << partition
+        idx = next_index(members_ids, idx)
+      end
+
+      partitions_per_member
+    end
+
+    def valid_sorted_partitions(members, partitions)
+      subscribed_topics = members.map do |id, metadata|
+        metadata && metadata.topics
+      end.flatten.compact
+
+      partitions
+        .select { |partition| subscribed_topics.include?(partition.topic) }
+        .sort_by { |partition| partition.topic }
+    end
+
+    def next_index(members_ids, idx)
+      idx += 1
+      idx = 0 if idx == members_ids.size
+
+      idx
+    end
```

Comment:
You can use this fancy Ruby trick instead, which I think is more readable:

```
irb(main):001:0> x = (0...5).cycle
irb(main):003:0> x.next
=> 0
irb(main):004:0> x.next
=> 1
irb(main):005:0> x.next
=> 2
irb(main):006:0> x.next
=> 3
irb(main):007:0> x.next
=> 4
irb(main):008:0> x.next
=> 0
```

Reply:
So you'd do e.g. this in `call`:

```ruby
idx = (0...members.size).cycle

# ...
id = idx.next
```

Reply:
oh nice will do!